File: plugin/src/main/java/spp/jetbrains/sourcemarker/VariableParser.java
package spp.jetbrains.sourcemarker;

import com.intellij.openapi.util.Pair;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class VariableParser {

    public static final String PATTERN_CURLY_BRACES = "\\{|\\}";
    public static final String EMPTY = "";
    public static final String SPACE = " ";
    public static final String DOLLAR = "$";

    public static Pair<Pattern, Pattern> createPattern(List<String> scopeVars) {
        Pattern variablePattern = null;
        Pattern templateVariablePattern = null;
        if (!scopeVars.isEmpty()) {
            StringBuilder sb = new StringBuilder("(");
            StringBuilder sbt = new StringBuilder("(");
            for (int i = 0; i < scopeVars.size(); i++) {
                sb.append("\\$").append(scopeVars.get(i));
                sbt.append("\\$\\{").append(scopeVars.get(i)).append("\\}");
                if (i + 1 < scopeVars.size()) {
                    sb.append("|");
                    sbt.append("|");
                }
            }
            sb.append(")(?:\\s|$)");
            sbt.append(")(?:|$)");
            variablePattern = Pattern.compile(sb.toString());
            templateVariablePattern = Pattern.compile(sbt.toString());
        }
        return Pair.create(variablePattern, templateVariablePattern);
    }

    public static Pair<String, List<String>> extractVariables(Pair<Pattern, Pattern> patternPair, String logPattern) {
        List<String> varMatches = new ArrayList<>();
        if (patternPair.first != null) {
            Matcher m = patternPair.first.matcher(logPattern);
            int matchLength = 0;
            while (m.find()) {
                String var = m.group(1);
                logPattern = logPattern.substring(0, m.start() - matchLength)
                        + logPattern.substring(m.start() - matchLength).replaceFirst(Pattern.quote(var), "{}");
                matchLength = matchLength + var.length() - 1;
                varMatches.add(var);
            }
        }

        if (patternPair.second != null) {
            Matcher m = patternPair.second.matcher(logPattern);
            int matchLength = 0;
            while (m.find()) {
                String var = m.group(1);
                logPattern = logPattern.substring(0, m.start() - matchLength)
                        + logPattern.substring(m.start() - matchLength).replaceFirst(Pattern.quote(var), "{}");
                matchLength = matchLength + var.length() - 1;
                var = var.replaceAll(PATTERN_CURLY_BRACES, EMPTY);
                varMatches.add(var);
            }
        }
        return Pair.create(logPattern, varMatches);
    }

    public static void matchVariables(Pair<Pattern, Pattern> patternPair, String input, Function<Matcher, Object> function) {
        if (patternPair.first != null) {
            Matcher match = patternPair.first.matcher(input);
            function.apply(match);
        }

        if (patternPair.second != null) {
            Matcher match = patternPair.second.matcher(input);
            function.apply(match);
        }
    }

    public static boolean isVariable(String text, String v) {
        String var = substringAfterLast(SPACE, text.toLowerCase());

        return (var.startsWith(DOLLAR) && !var.substring(1).equals(v)
                && v.toLowerCase().contains(var.substring(1)))
                || (var.startsWith("${") && !var.substring(2).equals(v)
                && v.toLowerCase().contains(var.substring(2)));
    }

    public static String substringAfterLast(String delimiter, String value) {
        int index = value.lastIndexOf(delimiter);
        if (index == -1) return value;
        else return value.substring(index + delimiter.length());
    }
}

File: plugin/src/main/kotlin/spp/jetbrains/sourcemarker/service/log/VariableParser.kt
/*
 * Source++, the open-source live coding platform.
 * Copyright (C) 2022 CodeBrig, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package spp.jetbrains.sourcemarker.service.log

import com.intellij.openapi.util.Pair
import java.util.function.Function
import java.util.regex.Matcher
import java.util.regex.Pattern

object VariableParser {

    const val EMPTY = ""
    const val SPACE = " "
    const val DOLLAR = "$"

    @JvmStatic
    fun createPattern(
        scopeVars: List<String>, varInitChar: String = DOLLAR,
        includeCurlyPattern: Boolean = true,
        ignoreCase: Boolean = false
    ): Pattern {
        val sb = StringBuilder("")
        if (scopeVars.isNotEmpty()) {
            sb.append("(")
            for (i in scopeVars.indices) {
                if (varInitChar.isNotEmpty()) {
                    sb.append("\\$varInitChar")
                }
                sb.append(scopeVars[i]).append("(?=\\s|$)")
                if (includeCurlyPattern) {
                    sb.append("|")
                    sb.append("\\$varInitChar\\{").append(scopeVars[i]).append("\\}")
                }
                if (i + 1 < scopeVars.size) {
                    sb.append("|")
                }
            }
            sb.append(")")
        }

        return if (ignoreCase) {
            Pattern.compile(sb.toString(), Pattern.CASE_INSENSITIVE)
        } else {
            Pattern.compile(sb.toString())
        }
    }

    @JvmStatic
    fun extractVariables(pattern: Pattern, logText: String): Pair<String, List<String>> {
        var logTemplate = logText
        val varMatches: MutableList<String> = ArrayList()
        if (pattern.pattern().isNotEmpty()) {
            val m = pattern.matcher(logTemplate)
            var matchLength = 0
            while (m.find()) {
                val variable = m.group(1)
                logTemplate = (logTemplate.substring(0, m.start() - matchLength)
                        + logTemplate.substring(m.start() - matchLength)
                    .replaceFirst(Pattern.quote(variable).toRegex(), "{}"))
                matchLength = matchLength + variable.length - 1

                if (variable.startsWith("$DOLLAR{")) {
                    varMatches.add(variable.substring(2, variable.length - 1))
                } else {
                    varMatches.add(variable.substring(1))
                }
            }
        }
        return Pair.create(logTemplate, varMatches)
    }

    fun matchVariables(pattern: Pattern, input: String, function: Function<Matcher, Any>) {
        val match = pattern.matcher(input)
        function.apply(match)
    }

    @JvmStatic
    fun isVariable(text: String, v: String): Boolean {
        val variable = text.toLowerCase().substringAfterLast(SPACE)
        return ((variable.startsWith(DOLLAR) && variable.substring(1) != v
                && v.toLowerCase().contains(variable.substring(1)))
                || (variable.startsWith("\${") && variable.substring(2) != v
                && v.toLowerCase().contains(variable.substring(2))))
    }
}
